### ISSUE TITLE
custom widget questions

### ISSUE THREAD

**Issue by bFlood** [eclipsesource/tabris-js#709]
1.  image property - in the docs it says "Accepts any value describing an image as outlined in the Tabris.js documentation. Other values are rejected. Will be passed to the native side as an object as described in the Tabris.js documentation." what is passed to the native side? base64 encoding of the image? an actual native object (e.g. Bitmap on Android). A sample would be great too

2.  javascript object - is it possible to use this as a property type? you can serialize to string and then deserialize in the widget native code but it would be nice to just pass in a js object and let the widget framework handle the process

**Comment by bFlood** [comment]
can an array of arrays be used as a property type? adding polylines and polygons to tabris maps and need to pass in an array of double[2] values as the coordinates

thanks

**Comment by bFlood** [comment]
in tabris-maps, how do you wire up simple properties on native classes? I can't seem to get the PropertyHandler class to work. for example, how would you get/set the Google Maps Marker title from the tabris-maps code?
https://github.com/eclipsesource/tabris-maps/blob/master/src/android/com/eclipsesource/tabris/maps/MarkerPropertyHandler.java#L29

the get/set properties on MarkerPropertyHandler are never called. the get/set methods on MarkerOperator are never called either. just a quick example would be very helpful

**Comment by mpost** [comment]
@bFlood For a simple implementation of a custom plugin you can look into this PropertyHandler: https://github.com/eclipsesource/tabris-calendar/blob/master/src/android/com/eclipsesource/tabris/calendar/CalendarWidgetPropertyHandler.java

**Comment by bFlood** [comment]
ok, I did look at that but it's defined differently vs MarkerPropertyHandler. which one is correct and should I just rewrite MarkerPropertyHandler? what are the pros/cons between the two different methods?

``` public class CalendarWidgetPropertyHandler extends TabrisWidgetPropertyHandler<CalendarView> {```
vs.
``` public class MarkerPropertyHandler<T extends Marker> implements IPropertyHandler<T> {```

should MarkerPropertyHandler be defined:
``` public class MarkerPropertyHandler extends TabrisWidgetPropertyHandler<Marker> {```

**Comment by mpost** [comment]
Both are correct... The MarkerPropertyHandler would allow you to be extend again while still maintaining the generic type.

**Comment by mpost** [comment]
While it was an internal decision to base the `MarkerPropertyHandler` on the `IPropertyHandler` interface we generally recommend to extend the `TabrisWidgetPropertyHandler`

**Comment by bFlood** [comment]
ok thanks, good to know

can you answer the three questions at the top of the post too?

**Comment by tbuschto** [comment]
image property: It passes on an object with the fields "src" and alternatively "scale", "width" and "height". 

javascript object: The property type "any" allows any JS value, including object. However, doing this with an object that contains functions is probably not a good idea.

"array" is a valid property type and allows any items inside the array.

**Comment by bFlood** [comment]
thanks @tbuschto 

properties - I've tried both way mentioned above, neither get called. I can change the marker properties directly after the MarkerOperator.java create call but anything coming through the tabris bridge do not get called

https://github.com/spatialdatalogic/tabris-maps/blob/master/src/android/com/eclipsesource/tabris/maps/MarkerPropertyHandler.java#L72

**Comment by mpost** [comment]
@bFlood Great to see you playing with this. :)

Your PropertyHandler has to implement the `TabrisPropertyHandler` interface and than your operator has to return that implementation in its `TabrisOperator.getPropertyHandler()`. It will not work with the internal `IPropertyHandler`. Sorry for the confusion.

We will update the tabris maps widget accordingly.

**Comment by jonek** [comment]
@bFlood Please take a look at my latest changes to the [tabris-maps](https://github.com/eclipsesource/tabris-maps) and [tabris-maps-example](https://github.com/eclipsesource/tabris-maps-example) projects. They should illustrate how to handle properties on map markers.
Feel free to submit pull requests on that projects if you continue to enhance the tabris-map custom widget.

**Comment by bFlood** [comment]
this is all working great now, thanks!

I'll definitely submit a PR when I get some more functionality added for Android. that said, I will probably fork and create a new repo that uses Google Maps on iOS (instead of Apple Maps)

loving tabris.js, incredible platform!

**Comment by mpost** [comment]
@bFlood Good to hear that you are getting along.

A Google Maps based Maps widget would be very interesting indeed. Keep us posted on your progress. :)

**Comment by bFlood** [comment]
added more Marker options, Polyline, Polygons and TileOverlays for android (see below). Still not sure the best way to pass a json object through the bridge, for now I'm serializing to string and then deserializing in the plugin (which works great but is not optimal if there is a lot of geometry data to put on the map)

1. @tbuschto how would I reference the JSONObject from the properties object if it were not passed in as a string (eg properties.getJSONObject("geometry") ????)
https://github.com/spatialdatalogic/tabris-maps/blob/master/src/android/com/eclipsesource/tabris/maps/PolygonOperator.java#L71


![screenshot_2015-12-23-15-52-41](https://cloud.githubusercontent.com/assets/404991/11984988/7336226a-a98e-11e5-910f-576c5d8515bd.png)

![screenshot_2015-12-23-15-53-32](https://cloud.githubusercontent.com/assets/404991/11984994/85b4961a-a98e-11e5-8b46-3147f0bdd23e.png)


![screenshot_2015-12-22-14-09-58](https://cloud.githubusercontent.com/assets/404991/11984997/8b5c92ac-a98e-11e5-9aa8-50e49f4555c0.png)

**Comment by mpost** [comment]
@bFlood These additions looks really nice. :)

Regarding your questions... As you pointed out there is no `properties.getJSONObject(..)` on the Properties object. While we see the occasional need for this scenario (we experienced it internally), Adding facilities to convert to object would be quite some work when doing it in a generic way. We used gson in our scenarios, similarly how you used the JSONObject.
